src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -660,42 +656,40 @@ impl Step for Miri {
         // # Run `cargo miri test`.
         // This is just a smoke test (Miri's own CI invokes this in a bunch of different ways and ensures
         // that we get the desired output), but that is sufficient to make sure that the libtest harness
-        // itself executes properly under Miri.
+        // itself executes properly under Miri, and that all the logic in `cargo-miri` does not explode.
+        // This is running the build `cargo-miri` for the given target, so we need the target compiler.
         let mut cargo = tool::prepare_tool_cargo(
             builder,
-            compiler,
-            Mode::ToolRustc,
-            host,
-            "run",
-            "src/tools/miri/cargo-miri",
+            target_compiler,
+            Mode::ToolStd, // it's unclear what to use here, we're not building anything just doing a smoke test!
```

Comment:
`ToolBootstrap` is the right choice if you don't build anything.

ref: https://github.com/rust-lang/rust/blob/40116ad1edef3ee66006e95fa9a483940aaa501c/src/bootstrap/src/core/build_steps/tool.rs#L79-L87

Reply:
That doesn't work for stages > 0 though.